### ISSUE TITLE
[BUG Table.readJson dropping valid JSON lines

### ISSUE THREAD

**Issue by andygrove** [rapidsai/cudf#14282]
**Describe the bug**
In the plugin, we have a ColumnVector containing JSON lines (concatenated together in one large string). Here is a sample:

```
GPU COLUMN combined - NC: 0 DATA: DeviceMemoryBufferView{address=0x30a013800, length=23082, id=-1} VAL: null
COLUMN combined - STRING
0 "{"teacher":null}
{"teacher":null}
...
{"teacher": "Qocwza","student": {"name": "Yiausu", "age": 19}}
{"teacher": "Rfwvcv","student": {"name": "Yxpbtq", "age": 12}}
{"teacher": "Spmydj","student": {"name": "Ggeyhv", "age": 16}}
{"teacher": "Jnbhwy","student": {"name": "Lmnkzw", "age": 19}}
{"teacher": "Gukkmo","student": {"name": "Xyyuy", "age": 19}}
...
```

We pass this to `Table.readJson` using the following code:

```scala
          val (names, rawTable) = withResource(combinedHost) { combinedHost =>
            val data = combinedHost.getData
            val start = combinedHost.getStartListOffset(0)
            val end = combinedHost.getEndListOffset(0)
            val length = end - start

            val jsonOptions = cudf.JSONOptions.builder().withRecoverWithNull(true).build()
            withResource(cudf.Table.readJSON(jsonOptions, data, start, length)) { tableWithMeta =>
              val names = tableWithMeta.getColumnNames
              (names, tableWithMeta.releaseTable())
            }
          }
```

The resulting table is missing some values. Note that the entry for `{"teacher": "Spmydj","student": {"name": "Ggeyhv", "age": 16}}` is `NULL` here.

```
DEBUG rawTable Table{columns=[ColumnVector{rows=512, type=STRING, nullCount=Optional.empty, offHeap=(ID: 260 7f4b7c3e1a00)}, ColumnVector{rows=512, type=STRUCT, nullCount=Optional.empty, offHeap=(ID: 261 7f4b7c3c1d60)}], cudfTable=139962183575680, rows=512}
GPU COLUMN 0 - NC: 228 DATA: DeviceMemoryBufferView{address=0x30a011a00, length=1696, id=-1} VAL: DeviceMemoryBufferView{address=0x30a006a00, length=64, id=-1}
COLUMN 0 - STRING
0 NULL
1 NULL
...
447 "Qocwza" 516f63777a61
448 "Rfwvcv" 526677766376
449 NULL
450 "Jnbhwy" 4a6e62687779
451 "Gukkmo" 47756b6b6d6f
```

**Steps/Code to reproduce bug**
The repro case is in https://github.com/NVIDIA/spark-rapids/pull/9423 in `test_from_json_struct_of_struct`.

**Expected behavior**
Data should not be dropped.

**Environment overview (please complete the following information)**
N/A

**Environment details**
N/A

**Additional context**

**Comment by elstehle** [comment]
Thanks for reporting the issue, @andygrove.

Unfortunately, I couldn't reproduce the issue. Could you share the full string you're trying to parse? It's surprising that line `449` is getting parsed to `NULL`, when its format is the same as for other lines that are getting parsed correctly. Could this have some other cause?

**Comment by andygrove** [comment]
I now have a repro case: https://github.com/rapidsai/cudf/pull/14291

**Comment by elstehle** [comment]
Just to follow up, @andygrove, I have finally gotten to the bottom of the issue. The issue only occurs when `withRecoverWithNull ` is used. The issue was introduced with the original PR that introduced that option https://github.com/rapidsai/cudf/pull/13344.

I'm currently elaborating options for resolving the issue and will likely have a resolution by the end of the week.

**Comment by elstehle** [comment]
@andygrove, I've put up https://github.com/rapidsai/cudf/pull/14309 to address this issue. Feel free to check if it properly addresses your issue.

**Comment by andygrove** [comment]
> @andygrove, I've put up #14309 to address this issue. Feel free to check if it properly addresses your issue.

Thank you @elstehle. I have confirmed that this resolves the issue.

**Comment by andygrove** [comment]
@elstehle I found one edge case where the last line will be dropped rather than replaced with null if it is invalid. This results in `Table.readJSON` returning one row less than the input data.

**Comment by elstehle** [comment]
Thanks for sharing, @andygrove! I'll look into it.

**Comment by elstehle** [comment]
> @elstehle I found one edge case where the last line will be dropped rather than replaced with null if it is invalid. This results in `Table.readJSON` returning one row less than the input data.

Thanks, Andy. I investigated the issue and it should only be an issue when the last line is both (a) incomplete, e.g., `{"a":` _and_ (b) does not end with a newline. I think the easiest way is to make sure that the JSON input is always terminated with a newline character. Is this something that could be done on the Spark side?

**Comment by elstehle** [comment]
Hi @andygrove!
This issue has been addressed by https://github.com/rapidsai/cudf/pull/14309 that was just merged. So, I'm closing this issue for now.

Please let me know if the following issue for incomplete last lines can be addressed by appending a newline on the Spark side. If that should turn out to not be feasible please let me know.

> @elstehle I found one edge case where the last line will be dropped rather than replaced with null if it is invalid. This results in Table.readJSON returning one row less than the input data.

**Comment by andygrove** [comment]
> Hi @andygrove! This issue has been addressed by #14309 that was just merged. So, I'm closing this issue for now.
> 
> Please let me know if the following issue for incomplete last lines can be addressed by appending a newline on the Spark side. If that should turn out to not be feasible please let me know.
> 
> > @elstehle I found one edge case where the last line will be dropped rather than replaced with null if it is invalid. This results in Table.readJSON returning one row less than the input data.

Thanks @elstehle. Yes, I have confirmed that adding a newline at the end resolves the issue for us. Thanks for looking into this.
